### PR TITLE
v2.6.1: Add LAN separate WAN modes and enhance installer mode detection and persistence with tests

### PR DESCRIPTION
**Pull request by jumpsmm7**
## Summary by CodeRabbit

* **New Features**
  * Added install-mode awareness across DNS/dnsmasq/IP set handling, including LAN bind-address refresh and WebUI port synchronization.
  * Improved DNS handoff and port-53 guarding with LAN/global bind “scope”, plus smarter dnsmasq restart behavior.
  * LAN mode now enforces IP set disablement and cleans related YAML entries.
* **Bug Fixes**
  * Prevented unwanted LAN-mode DNS NVRAM updates, stubby interactions, and IP set/dnsmasq actions.
  * Hardened startup/shutdown recovery and post-failure handling with restart suppression support.
* **Tests**
  * Expanded LAN/WAN DNS, dnsmasq gating, scoped readiness, installer LAN-mode, IP set, and lifecycle recovery test coverage.

### CODE REVIEW

File: installer [hunk 1]
```diff
@@ -6185,9 +7009,16 @@ WAN_IPADDR="$(nvram get wan_ipaddr 2>/dev/null)"
...
 
 [ -z "${ROUTER_MODEL}" ] && PTXT "${ERROR} This is an unsupported router, sorry." && sleep 3s && exit 1
-[ "$(nvram get sw_mode)" != "1" ] && PTXT "${ERROR} You are not running in router mode, sorry." && sleep 3s && exit 1
+PREVIOUS_ADGUARD_INSTALL_MODE="$(conf_value ADGUARD_INSTALL_MODE 2>/dev/null)"
+adguard_install_mode_detect || {
+	sleep 3s
+	exit 1
+}
+if adguard_install_mode_migration_required "$@"; then
+	adguard_migrate_detected_install_mode "${PREVIOUS_ADGUARD_INSTALL_MODE}" || exit 1
```
[the remark this answers is not in the document]
Comment (jumpsmm7) [reply installer]: @codex resolve this P1 , ensure all checksums are updated, and update all regression test pathways

File: S99AdGuardHome [hunk 2]
```diff
@@ -76,6 +76,43 @@ PREARGS="env TZ=/etc/localtime GOGC=${GOGC} GOMAXPROCS=${GOMAXPROCS} GOMEMLIMIT=
...
+}
+
+agh_dnsmasq_running() {
+	pidof dnsmasq >/dev/null 2>&1
+}
+
+agh_dns_handoff_required() {
+	case "$(agh_install_mode)" in
+		lan) agh_dnsmasq_running ;;
+		*) return 0 ;;
```
[the remark this answers is not in the document]
Comment (jumpsmm7) [reply S99AdGuardHome:126]: @codex resolve this P1 , ensure all checksums are updated, and update all regression test pathways

File: installer [hunk 3]
```diff
@@ -4529,9 +4824,468 @@ setup_restore_previous_yaml() {
...
+	fi
+	if [ -z "${bridge_binds}" ] && ai_have_cmd route; then
+		bridge_binds="$(route 2>/dev/null | awk -v lan_addr="${NET_ADDR}" '
+			function private_ip(ip) {
+				return ip ~ /^(10|127)\./ || ip ~ /^192\.168\./ || ip ~ /^172\.(1[6-9]|2[0-9]|3[0-1])\./
+			}
+			function router_ip(ip) {
+				split(ip, octets, ".")
+				if (octets[1] != "" && octets[2] != "" && octets[3] != "") { return octets[1] "." octets[2] "." octets[3] ".1" }
+				return ""
```
[the remark this answers is not in the document]
Comment (jumpsmm7) [reply installer]: @codex resolve this P2 , ensure all checksums are updated, and update all regression test pathways

File: installer [hunk 4]
```diff
@@ -4529,9 +4824,468 @@ setup_restore_previous_yaml() {
...
+		in_binds && $0 ~ /^[[:space:]]*($|#)/ { next }
+		in_binds && indentation($0) >= bind_indent && $0 ~ /^[[:space:]]*-[[:space:]]/ { next }
+		in_binds && indentation($0) > bind_indent { next }
+		in_binds { in_binds = 0 }
+		in_dns && (yaml_key_is($0, "upstream_dns") || yaml_key_is($0, "local_ptr_upstreams")) && !yaml_mapping_header_is($0, "upstream_dns") && !yaml_mapping_header_is($0, "local_ptr_upstreams") { exit 1 }
+		in_dns && (yaml_mapping_header_is($0, "upstream_dns") || yaml_mapping_header_is($0, "local_ptr_upstreams")) {
+			reverse_indent = indentation($0); in_reverse_upstreams = 1; print; next
+		}
+		in_reverse_upstreams && indentation($0) <= reverse_indent && $0 !~ /^[[:space:]]*($|#)/ { in_reverse_upstreams = 0 }
+		in_reverse_upstreams && old_reverse != "" { $0 = replace_reverse_endpoint($0, old_reverse, new_reverse) }
```
[the remark this answers is not in the document]
Comment (jumpsmm7) [reply installer]: @codex resolve this P2 , ensure all checksums are updated, and update all regression test pathways

File: installer [hunk 5]
```diff
@@ -4529,9 +4824,468 @@ setup_restore_previous_yaml() {
...
+				PTXT "${INFO} Legacy installation detected in LAN mode; updating mode-dependent AdGuardHome settings."
+			else
+				PTXT "${INFO} Router mode changed from ${previous_mode} to ${ADGUARD_INSTALL_MODE}; updating mode-dependent AdGuardHome settings."
+			fi
+			if ! setup_sync_mode_dependent_yaml_and_snapshot; then
+				PTXT "${ERROR} Unable to update the AdGuardHome configuration for ${ADGUARD_INSTALL_MODE} mode."
+				return 1
+			fi
+			adguard_install_feature_defaults || return 1
+			write_conf ADGUARD_NETCHECK_MODE "\"${ADGUARD_INSTALL_MODE}\"" || return 1
```
[the remark this answers is not in the document]
Comment (jumpsmm7) [reply installer]: @codex resolve this P1 , ensure all checksums are updated, and update all regression test pathways

File: AdGuardHome.sh [hunk 6]
```diff
@@ -787,6 +857,123 @@ interface_ipv6_addr() {
...
+	local BIND_HOSTS LAN_ADDR LAN_ADDR6 LAN_IF NVRAM_ADDR6 TEMP_FILE WEB_PORT
+	adguard_lan_mode || return 0
+	[ -f "${YAML_FILE}" ] || return 1
+	LAN_IF="$(nvram get lan_ifname 2>/dev/null)"
+	if [ -n "${LAN_IF}" ]; then
+		LAN_ADDR="$(interface_ipv4_addr "${LAN_IF}")"
+		LAN_ADDR6="$(interface_ipv6_addr "${LAN_IF}")"
+	fi
+	[ -n "${LAN_ADDR:-}" ] || LAN_ADDR="$(nvram get lan_ipaddr 2>/dev/null)"
+	[ -n "${LAN_ADDR}" ] || return 1
```
[the remark this answers is not in the document]
Comment (jumpsmm7) [reply AdGuardHome.sh]: @codex resolve this P1 , ensure all checksums are updated, and update all regression test pathways

File: installer [hunk 7]
```diff
@@ -4362,26 +4615,97 @@ inst_AdGuardHome() {
...
 		end_op_message 1 "$1"
 		return 1
 	fi
-	del_between_magic /jffs/scripts/service-event-end '# Asuswrt-Merlin-AdGuardHome-Installer'
-	if ! write_command_script /jffs/scripts/service-event-end 'if printf "%s" "$@" | /bin/grep -qE "^(((start|stop|restart|kill|reload)_?.*AdGuardHome)$)"; then { sh /jffs/addons/AdGuardHome.d/AdGuardHome.sh "$(printf "%s" "$@" | /bin/grep -oE "(start|stop|restart|kill|reload)")" x & }; fi # Asuswrt-Merlin-AdGuardHome-Installer'; then
-		PTXT "${ERROR} Required service-event-end hook could not be configured."
+	if ! adguard_migrate_detected_install_mode "${PREVIOUS_ADGUARD_INSTALL_MODE:-}"; then
 		adguard_restart_after_install_abort "${RESTART_AFTER_ABORT}"
 		end_op_message 1 "$1"
 		return 1
```
[the remark this answers is not in the document]
Comment (jumpsmm7) [reply installer]: @codex resolve this P2 , ensure all checksums are updated, and update all regression test pathways

**Comment by jumpsmm7** [comment]
@coderabbitai review

**Comment by jumpsmm7** [comment]
@codex review this code for bugs, security regression, and performance optimization

**Comment by jumpsmm7** [comment]
@codex address all feedback, update all checksums, and fix all regression test pathways

**Comment by jumpsmm7** [comment]
@codex review focus on edge cases, performance optimizations, security issues, security vulnerabilities, security concerns, behavioral regressions, missing tests, improper code behaviors

**Comment by jumpsmm7** [comment]
@codex address all feedback, update all checksums, and fix all regression test pathways

**Comment by jumpsmm7** [comment]
@codex review focus on edge cases, performance optimizations, security issues, security vulnerabilities, security concerns, behavioral regressions, missing tests, improper code behaviors

**Comment by jumpsmm7** [comment]
@codex review this code

**Comment by jumpsmm7** [comment]
@codex address all feedback, update all checksums, and fix all regression test pathways

File: installer [hunk 8]
```diff
@@ -6257,6 +7192,11 @@ esac
 
 case "${1:-}" in
 	install | update | backup | restore | doctor | ipset | netcheck | dns-port-policy | performance | migrate-runtime-defaults | uninstall)
+		case "${1:-}" in
+			install | update)
+				adguard_install_mode_detect || exit 1
+				;;
```
[the remark this answers is not in the document]
Comment (jumpsmm7) [reply installer]: @codex resolve this P1, update all checksums, and fix all regression test pathways

File: installer [hunk 9]
```diff
@@ -4529,9 +4862,546 @@ setup_restore_previous_yaml() {
...
+setup_sync_webui_port() {
+	local NEW_PORT CURRENT_WEB_LINE CURRENT_WEB_ADDRESS CURRENT_WEB_HOST CURRENT_WEB_PORT NEW_WEB_ADDRESS YAML_SYNC_TMP_FILE
+
+	NEW_PORT="$1"
+	port_is_valid "${NEW_PORT}" || return 1
+	CURRENT_WEB_LINE="$(awk '
+		/^http:[[:space:]]*$/ { in_http = 1; next }
+		in_http && /^[^[:space:]]/ { exit }
+		in_http && /^[[:space:]]*address:[[:space:]]*/ { print; exit }
+	' "${YAML_FILE}")"
```
[the remark this answers is not in the document]
Comment (jumpsmm7) [reply installer:5183]: @codex resolve this P2, update all checksums, and fix all regression test pathways

File: installer [hunk 10]
```diff
@@ -4360,25 +4611,107 @@ inst_AdGuardHome() {
...
-		return 1
-	fi
-	ptxt_ok "Asuswrt-Merlin event scripts configured."
-	if nvram get rc_support | grep -q 'mtlancfg'; then del_jffs_script /jffs/scripts/dnsmasq-sdn.postconf; fi
-	if nvram get rc_support | grep -q 'mtlancfg' && ! write_manager_script /jffs/scripts/dnsmasq-sdn.postconf 'dnsmasq-sdn $2'; then
-		PTXT "${ERROR} Required SDN dnsmasq event script could not be configured."
+	# Run mode-dependent YAML migration before this orchestration changes
+	# firewall state or event hooks, preserving the previous mode on failure.
+	if ! adguard_migrate_detected_install_mode "${PREVIOUS_ADGUARD_INSTALL_MODE:-}"; then
 		adguard_restart_after_install_abort "${RESTART_AFTER_ABORT}"
```
[the remark this answers is not in the document]
Comment (jumpsmm7) [reply installer]: @codex resolve this P2, update all checksums, and fix all regression test pathways

File: AdGuardHome.sh [hunk 11]
```diff
@@ -787,6 +874,152 @@ interface_ipv6_addr() {
...
+			"" | ::) ;;
+			*:*)
+				LAN_ADDR6="$(ip -o -6 addr list "${LAN_IF}" 2>/dev/null | awk -v candidate="${NVRAM_ADDR6}" '{ split($4, ip_addr, "/"); if (ip_addr[1] == candidate) { print candidate; exit } }')"
+				;;
+		esac
+	fi
+	BIND_HOSTS="$({
+		printf '%s\n' 127.0.0.1 "${LAN_ADDR}" "${LAN_ADDR6:-}"
+		private_ipv4_bridge_dns_options | awk 'NF > 1 { print $2 }'
+	} | awk 'NF && !seen[$0]++ { print }')"
```
[the remark this answers is not in the document]
Comment (jumpsmm7) [reply AdGuardHome.sh:917]: @codex resolve this P2, update all checksums, and fix all regression test pathways

**Comment by jumpsmm7** [comment]
@codex full review
